### PR TITLE
Update type hints

### LINKED ISSUE #754: AutoQASM: Use builtins for type hints now that Python 3.9 is min supported version

### PR DESCRIPTION
**Pull request by laurencap**
*Issue #, if available:* Closes #754

*Description of changes:*

Update type hints to use built-ins rather than `typing`. This is recommended for Python 3.9+.

This also allows us to remove a bit of code from `api.py` which was handling these different inputs, as AutoQASM needs to inspect type hints.

*Testing done:*
no new functionality, updates handled by existing tests.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/experimental/autoqasm/api.py [hunk 1]
```diff
@@ -17,7 +17,7 @@
 import functools
 import inspect
 from types import FunctionType
-from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+from typing import Any, Callable, Optional, Union
```
Comment (speller26) [inline src/braket/experimental/autoqasm/api.py:20]: You can also use the non-`typing` `Callable` now:
```python
from collections.abc import Callable
```

File: src/braket/experimental/autoqasm/program/program.py [hunk 2]
```diff
@@ -18,7 +18,7 @@
 import threading
 from dataclasses import dataclass
 from enum import Enum
-from typing import Any, Callable, Iterable, List, Optional, Union
+from typing import Any, Callable, Iterable, Optional, Union
```
Comment (speller26) [inline src/braket/experimental/autoqasm/program/program.py:21]:
```python
from collections.abc import Callable, Iterable
```

File: src/braket/experimental/autoqasm/transpiler/transpiler.py [hunk 3]
```diff
@@ -21,7 +21,7 @@
 import functools
 import importlib
 import inspect
-from typing import Any, Callable, Optional, Tuple, Union
+from typing import Any, Callable, Optional, Union
```
Comment (speller26) [inline src/braket/experimental/autoqasm/transpiler/transpiler.py:24]: ditto

File: src/braket/experimental/autoqasm/api.py (hunk 1 above)
Comment (laurencap) [reply src/braket/experimental/autoqasm/api.py:20]: Okay cool! I had also had a version of this that removed Option and Union before realizing that is only 3.10+ 😅

**Review by rmshaffer — APPROVED** [review]

**Pull request merged**